### PR TITLE
Disable keep-alive for cloud-event connections

### LINKED ISSUE #3190: Cloud Events delivery leaks open connections
# Expected Behavior
The pipeline controller does not leak open connection.
Possible fixes:
- re-use the HTTP client across reconciles
- do not keep idle connections open
- a combination of the two

# Actual Behavior
As described in https://github.com/tektoncd/triggers/issues/687, the pipeline controller seems to create a new HTTP client on every reconcile. The transport associated to it keeps the connections open.

# Steps to Reproduce the Problem

1. This is reproducible in the upstream CI/CD based Tekton (dogfooding)
2. Configure pipeline to send cloud events to an event listener sink
3. Run many taskruns and pipelineruns

# Additional Info
See https://github.com/tektoncd/triggers/issues/687 for more details

- Kubernetes version:

```
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.13-gke.1", GitCommit:"688c6543aa4b285355723f100302d80431e411cc", GitTreeState:"clean", BuildDate:"2020-07-21T02:37:26Z", GoVersion:"go1.13.9b4", Compiler:"gc", Platform:"linux/amd64"}
```

- Tekton Pipeline version:

```
Pipeline version: v0.15.2
Triggers version: v0.7.0
```

### PR DESCRIPTION
**Pull request by afrittoli**
# Changes

Disable keep alive forces the HTTP client to drop the connection
once a response is received. This avoids building up large numbers
of idle connections and it fixes the immediate issue.

After this we may want to see how to ensure we can re-use connection
and also set and idle-connection timeout.

Fixes #3190

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [n/a] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [n/a] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)
- [x] Release notes block has been filled in or deleted (only if no user facing changes)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes
```release-note
This release fixes an issue with the cloud event delivery that lead to an increasingly high number of open (idle) connections towards the configured target URL. The fix is achieved by disabling keep-alive, which means that - at least for now - we will be setting up a new connection for every cloud event. This is the current behaviour too, only now old connections will be closed immediatly.
```

/kind bug

### CODE REVIEW

File: pkg/reconciler/events/cloudevent/cloudeventclient.go [hunk 1]
```diff
@@ -35,11 +36,13 @@ type CECKey struct{}
 func withCloudEventClient(ctx context.Context, cfg *rest.Config) context.Context {
 	logger := logging.FromContext(ctx)
 
-	p, err := cloudevents.NewHTTP()
+	var useOnceTransport http.RoundTripper = &http.Transport{
```
Comment (ImJasonH) [inline pkg/reconciler/events/cloudevent/cloudeventclient.go:39]: Can you add a comment describing why we're doing this, with a link to the bug and a TODO to remove it when some upstream bug is closed? I don't want to forget we're doing this.

**Comment by dibyom** [comment]
This seems fine for now. We were already opening new connections for each cloud event delivery. This ensures that we close those connections instead of keeping them idle. In the future, would be nice to actually reuse the connections.
/lgtm

**Review by vdemeester — COMMENTED** [review]
/lgtm

File: pkg/reconciler/events/cloudevent/cloudeventclient.go (hunk 1 above)
Comment (afrittoli) [reply pkg/reconciler/events/cloudevent/cloudeventclient.go:39]: Good point, I will do.
I don't have an upstream bug created yet, but I'll link the pipeline bug for context.

**Comment by afrittoli** [comment]
Follow up feature: https://github.com/tektoncd/pipeline/issues/3204

**Comment by bobcatfish** [comment]
Cherry picking into 0.16.x in https://github.com/tektoncd/pipeline/pull/3215

**Pull request merged**